pkg/channelmetrics/metrics_collector/prometheus/collector.go

```diff
@@ -0,0 +1,111 @@
+package prometheus
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// MetricsCollector implements the |channelmetrics.MetricsCollector| interface using Prometheus.
+// It records various metrics related to channel operations.
+type MetricsCollector struct {
+	produceDuration prometheus.Histogram
+	consumeDuration prometheus.Histogram
+	channelLen      prometheus.Gauge
+	channelCap      prometheus.Gauge
+}
+
+var (
+	collectors   = make(map[string]*MetricsCollector)
+	collectorsMu sync.Mutex
+)
+
+// NewMetricsCollector creates a new MetricsCollector with
+// histograms for produce and consume durations, and gauges for channel length and capacity.
+// It accepts namespace, subsystem, and chanName parameters to organize metrics.
+// The function initializes and returns a pointer to a MetricsCollector struct
+// that contains the following Prometheus metrics:
+//
+//   - produceDuration: a Histogram metric that measures the duration of producing an item.
+//     It tracks the time taken to add an item to the ObservableChan.
+//     This metric helps to monitor the performance and latency of item production.
+//
+//   - consumeDuration: a Histogram metric that measures the duration of consuming an item.
+//     It tracks the time taken to retrieve an item from the ObservableChan.
+//     This metric helps to monitor the performance and latency of item consumption.
+//
+//   - channelLen: a Gauge metric that measures the current size of the channel buffer.
+//     It tracks the number of items in the channel buffer at any given time.
+//     This metric helps to monitor the utilization of the channel buffer.
+//
+//   - channelCap: a Gauge metric that measures the capacity of the channel buffer.
+//     It tracks the maximum number of items that the channel buffer can hold.
+//     This metric helps to understand the configuration and potential limits of the channel buffer.
+//
+// These metrics are useful for monitoring the performance and throughput of the ObservableChan.
+// By tracking the durations of item production and consumption, as well as the buffer size and capacity,
+// you can identify bottlenecks, optimize performance, and ensure that the ObservableChan is operating efficiently.
+func NewMetricsCollector(chanName, namespace, subsystem string) *MetricsCollector {
+	key := fmt.Sprintf("%s_%s_%s", namespace, subsystem, chanName)
+
+	collectorsMu.Lock()
+	defer collectorsMu.Unlock()
+
+	if collector, exists := collectors[key]; exists {
+		return collector
+	}
```

Comment:
Is there ever a case where we want `NewMetricsCollector` to return an existing collector? Or would this indicate we have a name collision?

Reply:
That would result in a collision, correct.

Reply:
Is a collision something that should never happen?